Add generic vector function material

GiudGiud opened the pull request:
Closes #18372

GiudGiud committed:
Add generic vector function material property, refs #18372

GiudGiud committed:
Add test kernel for vector material property diffusion, refs #18372

GiudGiud committed:
Fix up previous constructs, add a test for generic function material, refs #18372

Scale down from vector-mat-vector-diffusion to just vector-mat-diffusion

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/779157/) on 7b796b4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18373/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18373/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a969d327c03b526dd25c1af5b93f3b3aeabe2814`

GiudGiud committed:
Clang format

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/779183/) on f4fa0de wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18373/site)

This comment will be updated on new commits.

GiudGiud committed:
Fix up documentation for GenericFunctionVectorMaterial
Clang format

GiudGiud commented:
Well this is an interesting failure. Looks like there's something wrong with threads here.

GiudGiud commented:
ah the tensor maths is wrong is what is happening.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/892408/) on ba888b4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18373/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18373/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 44d8c685e45727d1bec41d4a56196a157d1a5e58`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/892467/) on 2124ab1 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>44d8c6</th>
      <th colspan=4>#18373 2124ab</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.95%</td>
      <td>81.96%</td>
      <td>+0.00%</td>
      <td>86.84%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72274</td>
      <td>72307</td>
      <td>+33</td>
      <td>33</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15915</td>
      <td>15920</td>
      <td>+5</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18373/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18373/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/18373/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 86.84% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented on framework/src/materials/GenericFunctionVectorMaterial.C:
```diff
@@ -0,0 +1,104 @@
+
+  if (_enable_stateful)
+  {
+    _properties_old.resize(num_names);
+    _properties_older.resize(num_names);
+  }
```
this is the same code as in GenericFunctionMaterial and it s all there is

I wonder if we re missing some code there?

GiudGiud replied:
it s deleted. no failures. So that was deprecated code.

andrsd commented on framework/include/materials/GenericFunctionVectorMaterial.h:
```diff
@@ -0,0 +1,61 @@
+
+#include "Material.h"
+
+// Forward Declarations
```
cpt. Obvious. - probably caused by copy-and-paste => remove this.

andrsd commented on framework/include/materials/GenericFunctionVectorMaterial.h:
```diff
@@ -0,0 +1,61 @@
+  std::vector<GenericMaterialProperty<RealVectorValue, is_ad> *> _properties;
+  std::vector<const MaterialProperty<RealVectorValue> *> _properties_old;
+  std::vector<const MaterialProperty<RealVectorValue> *> _properties_older;
+  std::vector<const Function *> _functions;
```
Missing doxygen for all the above member variables.

andrsd commented on framework/src/materials/GenericFunctionVectorMaterial.C:
```diff
@@ -0,0 +1,104 @@
+InputParameters
+GenericFunctionVectorMaterialTempl<is_ad>::validParams()
+{
+
```
Remove this empty line

andrsd commented on framework/src/materials/GenericFunctionVectorMaterial.C:
```diff
@@ -0,0 +1,104 @@
+  params.addDeprecatedParam<bool>("enable_stateful",
+                                  false,
+                                  "Enable the declaration of old and older values",
+                                  "all properties can implicitly become stateful");
```
Probably don't want to add a deprecated parameter in a new object. Recommended to remove and patch the code below not to use it.

GiudGiud committed:
Address David s review:
- remove deprecated stateful param in function materials
- remove line skips
- add doxygen
- remove captain obvious

GiudGiud commented:
Should be good now. I applied all your comments to GenericFunctionMaterial as well.
